### PR TITLE
Consolidate different functions into single get_params()

### PR DESCRIPTION
**Pull request by texasaggie97**
[X] This contribution adheres to [CONTRIBUTING.md]
### What does this Pull Request accomplish?
* Take the different flavors of get parameter list snippet and consolidate into single function
* Pick which flavor using enum
  ``` python
  class ParamListType(Enum):
      '''Type of parameter list to return'''
      API_METHOD = 1
      '''Used for methods param list for the public API'''
      IMPL_METHOD = 2
      '''Used for methods param list for implementation'''
      DISPLAY_METHOD = 3
      '''Used for methods param list for display (rst)'''
      LIBRARY_METHOD = 4
      '''Used for methods param list when calling library'''
      LIBRARY_CALL = 5
      '''Used for methods param list when calling into the DLL'''
      LIBRARY_CALL_TYPES = 6
      '''Used for methods param list types when calling into the DLL'''
  ```
* Add extra types to metadata
* Add session name to config metadata
* Each flavor has defaults for options:
  * skip_self (bool)
  * skip_session_handle (bool)
  * skip_output_parameters (bool)
  * skip_ivi_dance_size_parameter (bool)
  * session_name (str)
* Default options can be overridden by passing an additional dictionary
  ``` python
  get_params_snippet(f, helper.ParamListType.LIBRARY_CALL_TYPES, {'session_name': config['session_name']})
  ```
### Why should this Pull Request be merged?
* This will simplify parameter list generation when adding default parameters

### What testing has been done?
* Travis
* System tests

### CODE REVIEW

File: build/helper.py [hunk 1]
```diff
@@ -98,80 +102,170 @@ def normalize_string_type(d):
...
-            snippets.append(x['python_name'])
-    return ', '.join(snippets)
-
 
-def get_function_parameters_snippet(parameters, session_name='vi'):
-    '''Returns a string suitable for the parameter list of a method given a list of parameter objects
 
-    If session_name set, skip that parameter
+ParamListTypeDefaults = {}
+ParamListTypeDefaults[ParamListType.API_METHOD] = {
```
Comment (marcoskirsch) [inline build/helper.py]: Some of the things in this dictionary are "filtering/sorting" options, and some are "other" options. I think we should split it up.

And in general, I think we can break up this in two logical steps:

1. filter / reorder parameters
2. render the text snippet 

What do you think?

File: build/metadata_utilities.py [hunk 2]
```diff
@@ -248,7 +279,8 @@ def test_add_all_metadata_simple():
                         'mechanism': 'fixed',
                         'value': 1
                     },
-                    'type': 'ViSession'
+                    'type': 'ViSession',
+                    'library_call_name': 'self.vi',
```
Comment (marcoskirsch) [inline build/metadata_utilities.py]: What is this supposed to be? The name is not descriptive.

File: docs/nidmm/functions.rst [hunk 3]
```diff
@@ -1019,7 +1019,7 @@ nidmm.Session methods
     
 
 
-.. function:: fetch(maximum_time, reading)
+.. function:: fetch(maximum_time)
```
Comment (marcoskirsch) [inline docs/nidmm/functions.rst]: nice catch!

File: src/nifake/metadata/config.py [hunk 4]
```diff
@@ -5,6 +5,7 @@
     'c_function_prefix': 'niFake_',
     'driver_name': 'NI-FAKE',
     'session_description': 'An NI-FAKE session to a fake MI driver whose sole purpose is to test nimi-python code generation',
+    'session_name': 'vi',
```
Comment (marcoskirsch) [inline src/nifake/metadata/config.py]: Is this the name the generator will use for the member that stores the session handle?
Anything else?

File: src/nidmm/metadata/config.py [hunk 5]
```diff
@@ -5,6 +5,7 @@
     'c_function_prefix': 'niDMM_',
     'driver_name': 'NI-DMM',
     'session_description': 'An NI-DMM session to a National Instruments Digital Multimeter',
+    'session_name': 'vi',
```
Comment (marcoskirsch) [inline src/nidmm/metadata/config.py]: Suggested renames for clarity:

* session_description  -> session_class_description
* session_name -> session_handle_parameter_name

File: src/nifake/metadata/config.py (hunk 4 above)
Comment (marcoskirsch) [reply src/nifake/metadata/config.py]: It's not, it's the parameter name used by the C API as the session handle.

**Comment by marcoskirsch** [comment]
Right now in session.py we hard-coded the name of the Session member that holds the session handle "vi". This needs to match the new config key. It's a hidden contract. We should instead use the same config key so that this always matches.

File: build/helper.py (hunk 1 above)
Comment (texasaggie97) [reply build/helper.py]: Broken into two loops

File: build/metadata_utilities.py (hunk 2 above)
Comment (texasaggie97) [reply build/metadata_utilities.py]: This is the name used for the call into the library (DLL)

File: src/nidmm/metadata/config.py (hunk 5 above)
Comment (texasaggie97) [reply src/nidmm/metadata/config.py]: Changed

**Pull request merged**